src/compiler/crystal/semantic/call.cr

```diff
@@ -77,8 +77,13 @@ class Crystal::Call
 
     @target_defs = matches
 
-    bind_to matches if matches
-    bind_to block.break if block
+    # For a setter like `foo.x = y` we'd like to use `y`'s type
+    if self.setter?
+      bind_to args.first
```

Comment:
Note for later: it should probably be `bind_to args.last` instead to handle `[]=`